### ISSUE TITLE
Crash in 'BrowserSwitchClient' in version 3.14.0

### ISSUE THREAD

**Issue by sshropshire** [braintree/braintree-android-drop-in#196]
From: @TatyanaRTB

### General information

* SDK/Library version: Android SDK 4.0.1
* Environment: Development
* Android Version and Device: All Android versions with various devices
* Braintree dependencies: 

- com.braintreepayments.api:braintree:3.14.0
- com.braintreepayments.api:drop-in:4.6.0
- com.braintreepayments.api:google-payment:3.3.1

### Issue description

The issue is related to https://github.com/braintree/braintree_android/issues/308. Firstly we observed crash from the #308 and now after fixes crash happens in next step by code, when BrowserSwitchOptions.start is called:

ThreeDSecure:
```
public static void continuePerformVerification(BraintreeFragment fragment, ThreeDSecureRequest request, ThreeDSecureLookup threeDSecureLookup) {
    boolean showChallenge = threeDSecureLookup.getAcsUrl() != null;
    String threeDSecureVersion = threeDSecureLookup.getThreeDSecureVersion();
    fragment.sendAnalyticsEvent(String.format("three-d-secure.verification-flow.challenge-presented.%b", showChallenge));
    fragment.sendAnalyticsEvent(String.format("three-d-secure.verification-flow.3ds-version.%s", threeDSecureVersion));
    if (!showChallenge) {
        completeVerificationFlowWithNoncePayload(fragment, threeDSecureLookup.getCardNonce());
    } else if (!threeDSecureVersion.startsWith("2.")) {
        fragment.browserSwitch(13487, ThreeDSecureV1BrowserSwitchHelper.getUrl(fragment.getReturnUrlScheme(), fragment.getConfiguration().getAssetsUrl(), request, threeDSecureLookup));
    } else {
        performCardinalAuthentication(fragment, threeDSecureLookup);
    }
}
```
BrowserSwitchFragment:
```
public void browserSwitch(int requestCode, String url) {
    BrowserSwitchOptions browserSwitchOptions = (new BrowserSwitchOptions()).requestCode(requestCode).url(Uri.parse(url));
    this.browserSwitchClient.start(browserSwitchOptions, this);
}
```
BrowserSwitchClient:
```
public void start(BrowserSwitchOptions browserSwitchOptions, Fragment fragment) {
    if (fragment instanceof BrowserSwitchListener) {
        this.start(browserSwitchOptions, fragment, (BrowserSwitchListener)fragment);
    } else {
        throw new IllegalArgumentException("Fragment must implement BrowserSwitchListener.");
    }
}
```

The issue can be reproduced by the following steps:

1. Select credit card option in DropIn dialog
2. Enter card parameters
3. Press Go/Next button on the keyboard two-three times: https://monosnap.com/file/kcCFXR7wLZkKU7MtGu1bgvD6dHJaqF

Please block the button from several payment attempts if request to Braintree API is already in progress. Several payment flows are started in parallel and BraintreeFragment is detached from Activity when receive first successful payment result. Other payment results cause this crash.

```
java.lang.IllegalStateException: Fragment must be attached to an activity.
	at com.braintreepayments.browserswitch.BrowserSwitchClient.start(BrowserSwitchClient.java:206) ~[na:0.0]
	at com.braintreepayments.browserswitch.BrowserSwitchClient.start(BrowserSwitchClient.java:186) ~[na:0.0]
	at com.braintreepayments.browserswitch.BrowserSwitchFragment.browserSwitch(BrowserSwitchFragment.java:59) ~[na:0.0]
	at com.braintreepayments.api.ThreeDSecure.continuePerformVerification(ThreeDSecure.java:266) ~[na:0.0]
	at com.braintreepayments.api.ThreeDSecure$3.onLookupComplete(ThreeDSecure.java:155) ~[na:0.0]
	at com.braintreepayments.api.ThreeDSecure$7.success(ThreeDSecure.java:503) ~[na:0.0]
	at com.braintreepayments.api.internal.HttpClient$3.run(HttpClient.java:288) ~[na:0.0]
	at android.os.Handler.handleCallback(Handler.java:739) ~[na:0.0]
	at android.os.Handler.dispatchMessage(Handler.java:95) ~[na:0.0]
	at android.os.Looper.loop(Looper.java:157) ~[na:0.0]
	at android.app.ActivityThread.main(ActivityThread.java:5429) ~[na:0.0]
	at java.lang.reflect.Method.invoke(Native Method) ~[na:0.0]
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) ~[na:0.0]
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) ~[na:0.0]
```

**Comment by sshropshire** [comment]
@TatyanaRTB can you provide more information about what version of Android your emulator is using? I see the concert but unfortunately, I'm unable to reproduce at the moment. If possible, a stack trace or a video may help us to diagnose the issue and provide a fix more efficiently.

**Comment by TatyanaRTB** [comment]
@sshropshire Hello,

Here you can find log: 
[log_braintree_crash.txt](https://github.com/braintree/braintree-android-drop-in/files/5238837/log_braintree_crash.txt)

We hope it will be informative. 
Note that we can not reproduce the issue with Braintree Demo project.
The issue happens if press twice on this button: https://monosnap.com/file/kcCFXR7wLZkKU7MtGu1bgvD6dHJaqF
Affected different devices: Samsung S10e (Android 9), Google Pixel 3XL (Android 10), Huawei P9 EVA-L09 (Android 6)

Here is a fragment of out code: 
[log_braintree_crash.txt](https://github.com/braintree/braintree-android-drop-in/files/5238930/log_braintree_crash.txt)

Thank you in advance and hope to hear back from you soon.

**Comment by 0xAliHn** [comment]
We are also having the same issue during our testing even we are not using the **DropIn** dialog for our case.
Please see the log details. Found in device **OnePlus5T** 
```
Fatal Exception: java.lang.IllegalStateException: Fragment must be attached to an activity.
       at com.braintreepayments.browserswitch.BrowserSwitchClient.start(BrowserSwitchClient.java:206)
       at com.braintreepayments.browserswitch.BrowserSwitchClient.start(BrowserSwitchClient.java:186)
       at com.braintreepayments.browserswitch.BrowserSwitchFragment.browserSwitch(BrowserSwitchFragment.java:59)
       at com.braintreepayments.api.ThreeDSecure.continuePerformVerification(ThreeDSecure.java:266)
       at com.mytaxi.passenger.shared.braintree.ui.BraintreeScaActivity.continuePerformVerification(BraintreeScaActivity.java:60)
       at com.mytaxi.passenger.shared.braintree.ui.BraintreeScaActivityPresenter$lookupListener$1.onLookupComplete(BraintreeScaActivityPresenter.java:36)
       at com.braintreepayments.api.ThreeDSecure$7.success(ThreeDSecure.java:503)
       at com.braintreepayments.api.internal.HttpClient$3.run(HttpClient.java:288)
       at android.os.Handler.handleCallback(Handler.java:873)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:193)
       at android.app.ActivityThread.main(ActivityThread.java:6940)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

Occurring not that frequently but still its occurs pretty random. Any comments or fixes would be appreciated as soon as possible.

**Comment by sshropshire** [comment]
Hi @0xAliHn our browser switch library requires the Fragment to be attached to an Activity when called. You could defensively   guard against this scenario using the fragment's [isAdded()](https://developer.android.com/reference/android/app/Fragment.html#isAdded()) property.

**Comment by fradev94** [comment]
This issue has been solved? Because we have the same problem but rarely and for some devices for example Galaxy A51 (Android 10). The log is exactly the same

**Comment by TatyanaRTB** [comment]
@fradev94 @sshropshire Hello,

Issue is not reproduced for us starting from v5.0.2.

Thank you for evolution!

**Comment by fradev94** [comment]
@TatyanaRTB have you added something in your code? Or have you simply updated the drop in dependency in gradle?

**Comment by TatyanaRTB** [comment]
@fradev94 Only update the dependency for "com.braintreepayments.api:drop-in:$version"

![commit_braintree](https://user-images.githubusercontent.com/17588198/114006204-7160a500-988a-11eb-9b52-6965fc5af116.png)

**Comment by fradev94** [comment]
Thank you @TatyanaRTB for your answer. I also have the latest drop in lib version 5.2.0 but the crash happens sometimes. @sshropshire do you some advice for me please? Thank you

**Comment by sshropshire** [comment]
@fradev94 this may still be possible in an extreme edge case. Our advice for now is to assert that the Fragment's [isAdded()](https://developer.android.com/reference/android/app/Fragment.html#isAdded()) property is `true` before calling `continuePeformVerification()`.

**Comment by sshropshire** [comment]
@fradev94 as an aside, we do have a beta [version v4](https://github.com/braintree/braintree_android/blob/master/v4_MIGRATION_GUIDE.md) of the Braintree SDK that decouples base sdk functionality from Android `Fragment`. We plan on releasing an additional beta for Drop-In in the near future.

**Comment by fradev94** [comment]
Are there news about this crash please?

**Comment by sshropshire** [comment]
@fradev94 we will be releasing a stable version of `v4` very soon. I would suggest migrating to see if this fixes the issue. Otherwise, since `v3` is tightly coupled with Android `Fragment`, I would suggest guarding against `isAdded()` as mentioned previously.

**Comment by fradev94** [comment]
Hi, does the new dependency com.braintreepayments.api:drop-in:5.2.1 will solve this problem? Thank you

**Comment by sshropshire** [comment]
Hi @fradev94 we're working on migrating our DropIn library to `v4` of our Braintree Android SDK. We plan on doing a public beta with the updated SDK. We don't have a strict timeline, but we'll update this thread once the beta is available.

**Comment by sarahkoop** [comment]
Hi @fradev94 - we've released version `6.0.0-beta2` of the `drop-in` SDK that should resolve this issue.

**Comment by sarahkoop** [comment]
Version `6.0.0` of the Drop-in SDK has been released, which is compatible with v4 of the Braintree Android SDK and should resolve this issue.
